fix DS-17, DS-18, DS-19; .md file support

ItsNotJiaching opened the pull request:
The Slack Bot will now ask three questions and record the responses to the three questions when scraping a message. If a message has already been scraped, the bot will send a message notifying the user.

fix issues: DS-17, DS-18, DS-19

DS-17: The slack bot scraped all messages after it asked for a title, not just one.
Solution: implemented a variation of Izy's suggestion. The bot now checks if the timestamp of a parent thread has been scraped before asking the user to write a title, rather then using a timer to check, thus contributing to fixing DS-17.
In addition, if a message is not in direct response (in the same thread, right after the message sent by the bot) to the question the bot asked, the message will not be scraped, therefore also contributing to fixing DS-17

DS-18: update the slackbot to ask the user to upload a picture and then save the image link.
Solution: every time a reply message is sent to a parent message with a round_pushpin reaction, the bot checks whether it has asked for a title.  if it did, only then it will ask for a supplemental image. when the user submits a link to an image (a file will not work), it will record the image link and write it to the markdown file in .md format. if the user chooses not to upload an image, they will type "None" and the bot will record "No supplemental image was provided." This satisfies DS-18.

DS-19: update the slackbot to ask the user to add supplemental information and add it to the markdown file.
the same process used for DS-18 is applied here. after the bot asked and recorded the image link, it will ask the user to add supplemental text to the markdown file. if the user chooses not to add supplemental info, they will type "None" and the bot will not record any additional text. This satisfies DS-19.

Other changes:

1. updated the welcome message when a user joins the channel (or when "start" is typed) to describe how to use the bot.

2. stores the questions that the bot asks as variables in onboarding_tutorial.py

3. changed some imported libraries to remove redundancy and adjusted code accordingly

4. the bot will not listen to the event "pin_added" in my commits as I am only using message reactions.

5. updated code that displays the welcome message when "start" is typed to prevent error messages

6. added two functions.
the first checks whether a message has been scraped already.
the second writes text into the markdown file.

7. updated comments to explain updated code.

Current issues: 
No known bugs as of 8/16/22.

ItsNotJiaching committed:
fix DS-17, DS-18, DS-19; .md file support

The Slack Bot will now ask three questions and record the responses to the three questions when scraping a message. If a message has already been scraped, the bot will send a message notifying the user.

fix issues: DS-17, DS-18, DS-19

DS-17: The slack bot scraped all messages after it asked for a title, not just one.
Solution: implemented a variation of Izy's suggestion. The bot now checks if the timestamp of a parent thread has been scraped before asking the user to write a title, rather then using a timer to check, thus contributing to fixing DS-17.
In addition, if a message is not in direct response (in the same thread, right after the message sent by the bot) to the question the bot asked, the message will not be scraped, therefore also contributing to fixing DS-17

DS-18: update the slackbot to ask the user to upload a picture and then save the image link.
Solution: every time a reply message is sent to a parent message with a round_pushpin reaction, the bot checks whether it has asked for a title.  if it did, only then it will ask for a supplemental image. when the user submits a link to an image (a file will not work), it will record the image link and write it to the markdown file in .md format. if the user chooses not to upload an image, they will type "None" and the bot will record "No supplemental image was provided." This satisfies DS-18.

DS-19: update the slackbot to ask the user to add supplemental information and add it to the markdown file.
the same process used for DS-18 is applied here. after the bot asked and recorded the image link, it will ask the user to add supplemental text to the markdown file. if the user chooses not to add supplemental info, they will type "None" and the bot will not record any additional text. This satisfies DS-19.

Other changes:

1. updated the welcome message when a user joins the channel (or when "start" is typed) to describe how to use the bot.

2. stores the questions that the bot asks as variables in onboarding_tutorial.py

3. changed some imported libraries to remove redundancy and adjusted code accordingly

4. the bot will not listen to the event "pin_added" in my commits as I am only using message reactions.

5. updated code that displays the welcome message when "start" is typed to prevent error messages

6. added two functions.
the first checks whether a message has been scraped already.
the second writes text into the markdown file.

7. updated comments to explain updated code.

iengel35 commented on app.py:
```diff
@@ -59,79 +57,25 @@ def onboarding_message(event, client):
-    # Post the updated message in Slack
-    updated_message = client.chat_update(**message)
+        if not message_in_documentation(message_ts):
+            client.chat_postMessage(channel=channel_id, text=onboarding_tutorial.question_one, thread_ts=message_ts)
```
It's kind of misleading to pull your questions from the "onboarding_tutorial"- if you intend to use that file to hold all your bot responses (which I think makes sense!), I'd recommend renaming the file/variable to something more descriptive, like bot_messages and adding a comment to the header which explains what the purpose of the file is.

ItsNotJiaching replied:
Yes -- that makes sense. I did consider this but wasn't too confident in changing it up since I thought it may have been a little too chaotic, but I should've definitely done that when I felt it too.

the onboarding_tutorial.py file has been renamed to bot_messages.py.

iengel35 replied:
Thanks for updating!

iengel35 commented on app.py:
```diff
@@ -140,24 +84,95 @@ def message(event, client):
+    try:
+        if text.lower() == "start":
+            return start_onboarding(user_id, channel_id, client)
+    except AttributeError:
```
It's generally good practice to do something with your exception - why do AttributeErrors happen? What do they indicate? Should we log a warning and try again if we get one, or should we log an error message and quit the program? Something else?

ItsNotJiaching replied:
In hindsight I should've done that immediately when I found the error - I even had to go back to replicate how I got the error.

New code & comments have been added to explain what happens in the exception. 

I have a follow-up question: IF the error is minor & the comments explain the error fully, does the exception need to do anything? Or is it just fine leaving the exception code as None?

iengel35 replied:
Thanks for updating! Now it's much more clear what the error is and why it's ok to except it.

It's ideal to do something with the error, because errors are usually appearing for some reason/whatever is throwing the error isn't returning what you expect it to return. In cases like this, a simple log message is fine, but I would avoid just silently ignoring an error if possible.

Also, if you can build your explanation into your code/with logging, then it will always be clear to the user/developers what's going on, but comments won't be visible to the user.

ItsNotJiaching replied:
great! that makes sense. thanks for the explanation!

iengel35 commented on app.py:
```diff
@@ -140,24 +84,95 @@ def message(event, client):
+                    elif message["text"] == onboarding_tutorial.question_two:
+                        question_2_asked = True
+                    elif message["text"] == onboarding_tutorial.question_three:
+                        question_3_asked = True
```
You might consider using a single integer variable to track which question you're on, but up to you.

ItsNotJiaching replied:
Thanks - I didn't think of that. Using an integer variable definitely makes the code a lot more readable & is also easier to scale up if more questions were to be added.

A single integer variable has replaced the 3 Boolean variables to track what questions the bot has asked.

iengel35 reviewed (changes requested):
Left a few comments, but overall great progress! Excited to see the added functionality

ItsNotJiaching committed:
fix DS-17, DS-18, DS-19; .md file support (change #1)

Implemented Izy's requested changes.

1. renamed onboarding_tutorial.py to bot_messages.py to accurately represent the purpose of the file. A comment has also been added to clarify the file's purpose.

2. Every message the bot posts to Slack will be stored as variables in bot_messages.py to maintain consistency. These messages can be modified much easier in the future.

3. Code & comments have been added to explain what happens in the AttributeError on line 91/95.

4. A single integer variable has replaced the 3 Boolean variables to track what questions the bot has asked.

5. updated comments to explain updated code.

ItsNotJiaching commented:
ah oops - didn't realize hashtags referenced older PRs in GitHub

iengel35 commented on bot_messages.py:
```diff
@@ -1,4 +1,13 @@
+# questions that the bot asks when documenting
+question_one = "What do you want to title this wiki?"
+question_two = "Do you want to upload a supplemental image? If so, upload a link of the image. If not, type \"None\"."
+question_three = "Do you want to add anything else? If so, type any supplemental information you wish to add. If not, type \"None\"."
```
I should have caught this in the first review, but you could also make this an array indexed by your last_question_asked variable - totally no pressure to make this change in this PR though!

iengel35 reviewed (approved):
Suggested a possible change, but up to you on if you'd like to make it for this PR.
